Source/JavaScriptCore/bytecode/SpeculatedType.cpp

```diff
@@ -879,10 +884,15 @@ SpeculatedType typeOfDoubleAbs(SpeculatedType value)
 
 SpeculatedType typeOfDoubleRounding(SpeculatedType value)
 {
-    // Double Pure NaN can becomes impure when converted back from Float.
-    // and vice versa.
-    if (value & SpecDoubleNaN)
-        value |= SpecDoubleNaN;
+    if (isARM64()) {
+        if (value & SpecDoubleImpureNaN)
+            value |= SpecDoublePureNaN;
+    } else {
+        // Double Pure NaN can becomes impure when converted back from Float.
+        // and vice versa.
+        if (value & SpecDoubleNaN)
+            value |= SpecDoubleNaN;
```

Comment:
Is this correct?

Reply:
For non ARM64, we are taking a bit conservative (pure NaN may be impure NaN), but we can change it if it gets proven on each arch.